### PR TITLE
Scalable SC

### PR DESCRIPTION
**Pull request by ElusAegis**
Allows the SC to run with more than 275 puzzle parts by allowing to execute `initiate` and `setCommits` in the SC in Batch Size parts, that fit the gas limit.

### CODE REVIEW

File: src/contracts/SmartContract.sol [hunk 1]
```diff
@@ -29,31 +31,42 @@ contract SmartContract {
...
-        initialTimestamp = block.timestamp;
+        if (amountOfPuzzleParts == 0) {
+            startTime = _startTime;
+            helperID = _helperID;
+            initialTimestamp = block.timestamp;
+            owner = msg.sender;
+        }
 
         uint receivedValue = msg.value;
+        uint currentIndex = amountOfPuzzleParts;
```
Comment (danrr) [inline src/contracts/SmartContract.sol]: does this need to be a separate variable?

File: src/contracts/SmartContract.sol [hunk 2]
```diff
@@ -29,31 +31,42 @@ contract SmartContract {
     uint public extraTime;
     address public helperID;
     uint public initialTimestamp;
+    Status public contractStatus = Status.Setup;
 
     mapping(uint => PuzzlePart) public puzzleParts;
     uint amountOfPuzzleParts;
```
Comment (danrr) [inline src/contracts/SmartContract.sol]: Should this be initialised to 0 explicitly?

File: src/tlp_lib/smartcontracts/EthereumSC.py [hunk 3]
```diff
@@ -45,8 +45,21 @@ def commitments(self) -> TLP_Digests:
 
     @commitments.setter
     def commitments(self, commitments: TLP_Digests):
-        if not self._has_succeeded(self._contract.functions.setCommitments(commitments)):
-            raise RuntimeError("Commitments were not set correctly")
+        total_commitments = len(commitments)
+        start_index = 0
+
+        while start_index < total_commitments:
```
Comment (danrr) [inline src/tlp_lib/smartcontracts/EthereumSC.py]: I don't think you need to do this manually, there's a function in the `itertools` module that helps with this: https://docs.python.org/3.12/library/itertools.html#itertools.batched

You can then iterate over the generator like

```py
for commitments_batch in itertools.batched(commitments, self._SC_PUZZLE_BATCH_SIZE):
```

File: src/tlp_lib/smartcontracts/EthereumSC.py [hunk 4]
```diff
@@ -184,21 +199,61 @@ def _deploy_contract(
...
+        helper_id: int | ChecksumAddress,
+    ) -> None:
+        # Ensure that all lists have the same length
+        assert len(coins) == len(extra_times) == len(upper_bounds), "All input lists must have the same length"
+
+        total_parts = len(coins)
+        start_index = 0
+
+        while start_index < total_parts:
+            end_index = min(start_index + self._SC_PUZZLE_BATCH_SIZE, total_parts)
```
Comment (danrr) [inline src/tlp_lib/smartcontracts/EthereumSC.py]: ditto on slicing, but you can zip the batched lists, so you do

```py 
for coins_batch, extra_times_batch, upper_bounds_batch in zip(...)
```

File: src/tlp_lib/smartcontracts/EthereumSC.py [hunk 5]
```diff
@@ -184,21 +199,61 @@ def _deploy_contract(
...
+
+            # Create a slice for the current batch
+            coins_batch = coins[start_index:end_index]
+            extra_times_batch = extra_times[start_index:end_index]
+            upper_bounds_batch = upper_bounds[start_index:end_index]
+
+            # Calculate the value to send with this batch
+            value_to_send = functools.reduce(lambda x, y: x + y, coins_batch)
```
Comment (danrr) [inline src/tlp_lib/smartcontracts/EthereumSC.py]: `sum(coins_batch)`?

**Review by danrr — COMMENTED** [review]
This looks good on first reading, left a few minor comments. I'm going to try to run things and see how it works

File: src/contracts/SmartContract.sol (hunk 1 above)
Comment (ElusAegis) [reply src/contracts/SmartContract.sol]: It does improve the readability of the code in my opinion, though can be removed

File: src/contracts/SmartContract.sol (hunk 2 above)
Comment (ElusAegis) [reply src/contracts/SmartContract.sol]: No, but again, feels better for readability

File: src/contracts/SmartContract.sol (hunk 1 above)
Comment (danrr) [reply src/contracts/SmartContract.sol]: I get your point, though `currentIndex + i` is a tad confusing

File: src/tlp_lib/smartcontracts/EthereumSC.py (hunk 3 above)
Comment (ElusAegis) [reply src/tlp_lib/smartcontracts/EthereumSC.py]: Thank you, did not know this existed, I have changed it accordingly

Comment (danrr) [reply src/tlp_lib/smartcontracts/EthereumSC.py]: yeah, the standard library has been getting a lot of useful stuff with the last few releases

File: src/tlp_lib/smartcontracts/EthereumSC.py (hunk 5 above)
Comment (ElusAegis) [reply src/tlp_lib/smartcontracts/EthereumSC.py]: this is too simple 🙃😄

File: src/tlp_lib/smartcontracts/EthereumSC.py (hunk 4 above)
Comment (ElusAegis) [reply src/tlp_lib/smartcontracts/EthereumSC.py]: also changed this

File: src/contracts/SmartContract.sol (hunk 1 above)
Comment (ElusAegis) [reply src/contracts/SmartContract.sol]: Agree, changed it to `startIndex` to also be consistent with the `setCommitment`

**Pull request merged**